src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractElementNode.java

```diff
@@ -0,0 +1,348 @@
+                }
+                else {
+                    result = EmptyIterator.ofNodes();
+                }
+                break;
+            case AxisInfo.DESCENDANT_OR_SELF:
+                result = new DescendantIterator(this, DescendantIterator.StartWith.CURRENT_NODE);
+                break;
+            case AxisInfo.PARENT:
+                result = SingleNodeIterator.makeIterator(parent);
+                break;
+            case AxisInfo.SELF:
+                result = SingleNodeIterator.makeIterator(this);
+                break;
+            case AxisInfo.FOLLOWING_SIBLING:
+                result = getFollowingSiblingsIterator();
+                break;
+            case AxisInfo.PRECEDING_SIBLING:
+                result = getPrecedingSiblingsIterator();
+                break;
+            case AxisInfo.FOLLOWING:
+                result = new FollowingIterator(this);
+                break;
+            case AxisInfo.PRECEDING:
+                result = new PrecedingIterator(this);
+                break;
+            default:
+                throw throwUnsupportedOperationException();
+        }
+
+        return result;
+    }
+
+    /**
+     * Returns preceding sibling axis iterator.
+     *
+     * <p>Reason of suppression for resource, IOResourceOpenedButNotSafelyClosed:
+     * {@link AxisIterator} implements {@link java.io.Closeable} interface,
+     * but none of the subclasses of the {@link AxisIterator}
+     * class has non-empty {@code close()} method.
+     *
+     * @return iterator
+     */
+    private AxisIterator getPrecedingSiblingsIterator() {
+        final AxisIterator result;
+        if (indexAmongSiblings == 0) {
+            result = EmptyIterator.ofNodes();
+        }
+        else {
+            result = new ReverseListIterator(getPrecedingSiblings());
+        }
+        return result;
+    }
+
+    /**
+     * Returns following sibling axis iterator.
+     *
+     * <p>Reason of suppression for resource, IOResourceOpenedButNotSafelyClosed:
+     * {@link AxisIterator} implements {@link java.io.Closeable} interface,
+     * but none of the subclasses of the {@link AxisIterator}
+     * class has non-empty {@code close()} method.
+     *
+     * @return iterator
+     */
+    private AxisIterator getFollowingSiblingsIterator() {
+        final AxisIterator result;
+        if (indexAmongSiblings == parent.getChildren().size() - 1) {
+            result = EmptyIterator.ofNodes();
+        }
+        else {
+            result = new ArrayIterator.OfNodes<>(
+                    getFollowingSiblings().toArray(EMPTY_ABSTRACT_NODE_ARRAY));
+        }
+        return result;
+    }
+
+    /**
+     * Returns following siblings of the current node.
+     *
+     * @return siblings
+     */
+    private List<AbstractNode> getFollowingSiblings() {
+        final List<AbstractNode> siblings = parent.getChildren();
+        return siblings.subList(indexAmongSiblings + 1, siblings.size());
```

Comment:
Can we return an unmodifiable list here, and below?

Reply:
I was for this change, but the more I look at it, the more it isn't needed and this is because of how we use this method.

`getFollowingSiblings` is a `private` method and is only called by `getFollowingSiblingsIterator`. It usage inside `getFollowingSiblingsIterator` is to just convert it to an array. We would gain nothing by going through the extra step of making sure the list cannot be modified.

`getPrecedingSiblings` is similar, but it isn't converted to an array. It is passed to `ReverseListIterator`.

I can change that one as well, but I see the gain as very minimal since this is openly used in places.

Reply:
Sounds reasonable to just change `getPrecedingSiblings`, but it's up to you.

Reply:
Resolved.